### PR TITLE
Fix initial page startup race for late-loading client bundles

### LINKED ISSUE #3150: pageLifecycle misses startup during interactive if DOMContentLoaded already fired
Reported by @lesleh from @simplybusiness while upgrading to a recent v16 release.

Current `pageLifecycle` logic does this:

- `document.readyState === "complete"` → initialize immediately
- otherwise → wait for `DOMContentLoaded`

That behavior came from [PR #2295](https://github.com/shakacode/react_on_rails/pull/2295) to avoid the older deferred-script registration race from [#1150](https://github.com/shakacode/react_on_rails/issues/1150): if startup runs during `interactive`, deferred/module scripts may not have finished `ReactOnRails.register(...)` yet.

However, it regresses another startup path that earlier async-loading fixes had covered (see [#1330](https://github.com/shakacode/react_on_rails/issues/1330), [#1331](https://github.com/shakacode/react_on_rails/pull/1331), and [#1773](https://github.com/shakacode/react_on_rails/pull/1773)):

- the client bundle starts while `document.readyState === "interactive"`
- `DOMContentLoaded` has already fired
- `readyState` is not yet `"complete"`

In that window, React on Rails attaches the `DOMContentLoaded` listener too late, so startup never runs.

## Why this matters

There are two competing requirements:

1. We must **not** initialize immediately during `interactive`, or we can reintroduce the `#1150` / `#2295` race.
2. We must still recover if `DOMContentLoaded` was already missed before startup attached its listener.

A minimal local side-by-side repro demonstrates both cases:

### Missed `DOMContentLoaded` during `interactive`

- old `readyState !== 'loading'` behavior: `['setup']`
- current `readyState === 'complete'` behavior: `[]`
- desired hybrid behavior: `['readystatechange-complete-handler', 'setup']`

### Deferred/module registration race

- old `readyState !== 'loading'` behavior: `['hydrate-before-register:ERROR']`
- current `readyState === 'complete'` behavior: `['hydrate-after-register:OK']`
- desired hybrid behavior: `['hydrate-after-register:OK']`

## Expected behavior

- `loading` → wait for `DOMContentLoaded`
- `interactive` → wait for `DOMContentLoaded`, but if it was already missed, recover on `readystatechange => complete`
- `complete` → initialize immediately

This preserves the `#1150` / `#2295` safety while restoring the missed-startup case.

### PR DESCRIPTION
**Pull request by ihabadham**
### Summary
Fixes a client lifecycle gap where React on Rails can miss the initial page startup if the bundle begins executing after `DOMContentLoaded` has already fired but before the document reaches `complete`.

Fixes #3150

### Problem
Core startup currently initializes immediately only when `document.readyState === "complete"`, and otherwise waits for `DOMContentLoaded`.

That avoids starting too early during `interactive`, when deferred or module scripts may still be registering components.

But there is another browser timing window:

- the bundle starts while `document.readyState === "interactive"`
- `DOMContentLoaded` has already fired
- the document is not yet `"complete"`

In that case, the `DOMContentLoaded` listener is attached too late, so the initial page-load callbacks never run.

### What changed
Adds a `readystatechange → complete` fallback listener when registration happens during `interactive`. Both listeners share a single handler that removes both on first fire, so initialization runs exactly once in real browsers.

### Test plan
- Added regression coverage for the missed-DOMContentLoaded-during-interactive scenario. Verified by temporarily reverting the fix and running the new tests against main's `pageLifecycle.ts` — exactly the new tests failed.
- Verified initial startup now runs when the client bundle begins during `interactive` after `DOMContentLoaded` has already fired.

### CODE REVIEW

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 1]
```diff
@@ -28,34 +32,88 @@ function runPageUnloadedCallbacks(): void {
...
+    return false;
+  }
+
+  if (document.readyState === 'complete') {
+    cleanupInitialPageLoadListeners();
+    ensureNavigationListenersInstalled();
+  }
+
+  runPageLoadedCallbacks();
+  return true;
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts]: Superseded — `consumeInitialPageLoadIfNeeded` was removed by the revert in `05e43fb3a` and the simplification in `3c3ebcf07`. Current code uses a single shared handler (`pageLifecycle.ts:82-97`) with no asymmetry between interactive and loading paths.

File: packages/react-on-rails/src/clientStartup.ts [hunk 2]
```diff
@@ -1,31 +1,46 @@
...
-  onPageLoaded(reactOnRailsPageLoaded);
+  onPageLoaded(runAutomaticPageLoad);
+  return true;
+}
+
+export function reactOnRailsPageLoaded() {
+  const startupWasPending = clientStartup();
+  if (startupWasPending) {
+    consumeInitialPageLoadIfNeeded();
+    return;
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/clientStartup.ts]: Superseded — `startupWasPending` was part of commit `219545c9c`, reverted in `05e43fb3a`. Current `reactOnRailsPageLoaded` (clientStartup.ts:7) just calls `renderAllComponents()`; no third-case branching.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 3]
```diff
@@ -28,34 +32,88 @@ function runPageUnloadedCallbacks(): void {
...
 }
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts]: Superseded — same reason as https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3088617736. `consumeInitialPageLoadIfNeeded` no longer exists.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 4]
```diff
@@ -67,20 +67,51 @@ function initializePageEventListeners(): void {
...
+  // fallback page-load sweep. See pageLifecycle.test.js for regression coverage of both cases.
   if (document.readyState === 'complete') {
     setupPageNavigationListeners();
   } else {
-    document.addEventListener('DOMContentLoaded', setupPageNavigationListeners);
+    let isSetupComplete = false;
+    let setupPageNavigationListenersOnce: () => void;
+    let setupPageNavigationListenersOnComplete: () => void;
+
+    const cleanupListeners = (): void => {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts]: Superseded — the forward-declared `let` bindings and `cleanupListeners` were removed in the simplification commit `3c3ebcf07`. The current handler (`pageLifecycle.ts:85-90`) has no TDZ-adjacent pattern.

File: packages/react-on-rails/src/clientStartup.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/clientStartup.ts]: Superseded — `clientStartup()` is no longer called from `reactOnRailsPageLoaded()`. That change was in commit `219545c9c`, reverted in `05e43fb3a`. `setOptions({ turbo: true })` timing is unchanged from before this PR.

File: packages/react-on-rails/tests/clientStartup.test.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    document.body.innerHTML = '';
+    addEventListenerSpy.mockRestore();
+    delete globalThis.ReactOnRails;
+    // eslint-disable-next-line no-underscore-dangle
+    delete globalThis.__REACT_ON_RAILS_EVENT_HANDLERS_RAN_ONCE__;
+  });
+
+  it('does not render twice when manual page load runs during interactive before deferred startup registration', async () => {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/clientStartup.test.js]: Out of scope — this PR doesn't modify `clientStartup.test.js`. Branch scope: `CHANGELOG.md`, `pageLifecycle.ts`, `pageLifecycle.test.js`.

File: packages/react-on-rails/tests/clientStartup.test.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    expect(counts).toEqual({ storeCalls: 1, rendererCalls: 1 });
+
+    await waitForNextTick();
+    expect(counts).toEqual({ storeCalls: 1, rendererCalls: 1 });
+
+    setReadyState('complete');
+    document.dispatchEvent(new Event('readystatechange'));
+    await waitForNextTick();
+
+    expect(counts).toEqual({ storeCalls: 1, rendererCalls: 1 });
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/clientStartup.test.js]: Out of scope — same as https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3088620114. `clientStartup.test.js` isn't modified by this PR.

File: packages/react-on-rails/tests/pageLifecycle.test.js [hunk 7]
```diff
@@ -92,6 +95,7 @@ describe('pageLifecycle', () => {
     expect(callback).not.toHaveBeenCalled();
     // Verify that a DOMContentLoaded listener was added when readyState is 'interactive'
     expect(addEventListenerSpy).toHaveBeenCalledWith('DOMContentLoaded', expect.any(Function));
+    expect(addEventListenerSpy).toHaveBeenCalledWith('readystatechange', expect.any(Function));
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:98]: Declining — test-style preference. The positive assertion in the interactive test combined with the `if (document.readyState === 'interactive')` guard at `pageLifecycle.ts:94` already pins the asymmetry. Adding a symmetric negative assertion would duplicate coverage without catching a likely regression path.

[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:98]: Duplicate of https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3089106403.

File: packages/react-on-rails/tests/pageLifecycle.test.js [hunk 8]
```diff
@@ -31,6 +31,9 @@ describe('pageLifecycle', () => {
   // eslint-disable-next-line global-require
   const importPageLifecycle = () => require('../src/pageLifecycle.ts');
 
+  const getEventHandler = (eventName) =>
+    addEventListenerSpy.mock.calls.find((call) => call[0] === eventName)?.[1];
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:35]: Declining — tests don't register the same event name multiple times within a single test and modules reset between tests. If that invariant changes later, the helper can be updated alongside the test that introduces the multi-registration pattern.

[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:35]: Duplicate of https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3089107610.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 9]
```diff
@@ -67,20 +67,33 @@ function initializePageEventListeners(): void {
...
+  // fallback page-load sweep. See pageLifecycle.test.js for regression coverage of both cases.
   if (document.readyState === 'complete') {
     setupPageNavigationListeners();
   } else {
-    document.addEventListener('DOMContentLoaded', setupPageNavigationListeners);
+    const initialPageLoadHandler = (event: Event): void => {
+      if (event.type === 'readystatechange' && document.readyState !== 'complete') return;
+      document.removeEventListener('DOMContentLoaded', initialPageLoadHandler);
+      document.removeEventListener('readystatechange', initialPageLoadHandler);
+      setupPageNavigationListeners();
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts:89]: Declining the defensive flag. Idempotency of `setupPageNavigationListeners` is enforced by the `removeEventListener` calls in the initial-page-load handler. Adding an `isNavigationListenersSetUp` flag inside the callee would mask the "cleanup was accidentally deleted" bug class rather than surface it; we want the safety invariant to live at the caller site where it's visible.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 10]
```diff
@@ -28,34 +32,88 @@ function runPageUnloadedCallbacks(): void {
...
+  if (initialPageLoadReadyHandler) {
+    document.removeEventListener('DOMContentLoaded', initialPageLoadReadyHandler);
+    initialPageLoadReadyHandler = null;
+  }
+
+  if (initialPageLoadCompleteHandler) {
+    document.removeEventListener('readystatechange', initialPageLoadCompleteHandler);
+    initialPageLoadCompleteHandler = null;
+  }
+}
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts]: Declining — as noted in the review, this is a harmless no-op. Adding a `readyState === 'interactive'` guard around `removeEventListener('readystatechange', ...)` would duplicate the guard already present at the `addEventListener` site with no behavioral change.

File: packages/react-on-rails/tests/pageLifecycle.test.js [hunk 11]
```diff
@@ -356,18 +362,38 @@ describe('pageLifecycle', () => {
...
+      expect(domContentLoadedHandler).toBeDefined();
+
+      domContentLoadedHandler(new Event('DOMContentLoaded'));
+      expect(callback).toHaveBeenCalledTimes(1);
+
+      // The handler's cleanup is what prevents the later readystatechange→complete
+      // event from firing the callback a second time in real browsers.
+      expect(removeEventListenerSpy).toHaveBeenCalledWith('DOMContentLoaded', expect.any(Function));
+      expect(removeEventListenerSpy).toHaveBeenCalledWith('readystatechange', expect.any(Function));
+    });
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js]: The test this was against was removed in `2a655fd19`. The `removeEventListenerSpy.mockRestore()` approach was considered — it doesn't actually work in this harness: tests retrieve and invoke handlers via the spy's captured `addEventListener` calls, and those handler references remain callable after `removeEventListener` unhooks them from the document. A true behavioral test would require replacing the spy-based dispatch harness, which is out of scope.

File: packages/react-on-rails/tests/pageLifecycle.test.js [hunk 12]
```diff
@@ -356,17 +362,17 @@ describe('pageLifecycle', () => {
...
+      expect(readyStateChangeHandler).toBeDefined();
 
-      // Simulate state transition: deferred scripts complete, then DOMContentLoaded fires
       setReadyState('complete');
-      domContentLoadedHandler();
+      readyStateChangeHandler(new Event('readystatechange'));
 
-      // Now callback should have been called
+      // Now callback should have been called through the complete fallback
       expect(callback).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:376]: Declining — same reasoning as why the analogous test was dropped in `2a655fd19`. Asserting `removeEventListener` was called is a mechanism-level assertion, not a behavioral one. It would fail on valid refactors (e.g. AbortController) that preserve the cleanup invariant via a different mechanism, without catching cleanup bugs that leave the calls in place but misapply them. The invariant itself (`setupPageNavigationListeners` runs at most once) can't be tested behaviorally in this harness.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 13]
```diff
@@ -67,20 +125,36 @@ function initializePageEventListeners(): void {
...
+      if (document.readyState === 'complete') {
+        handleAutomaticInitialPageLoad();
+      }
+    };
+
+    document.addEventListener('DOMContentLoaded', initialPageLoadReadyHandler);
+
+    if (document.readyState === 'interactive') {
+      document.addEventListener('readystatechange', initialPageLoadCompleteHandler);
+    }
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts]: The rationale is already documented in the block comment at `pageLifecycle.ts:70-81`, which covers why the `interactive` path needs the `complete` fallback. Adding another comment at the registration site would duplicate that explanation.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 14]
```diff
@@ -67,20 +67,36 @@ function initializePageEventListeners(): void {
...
+      // readystatechange→complete transition would run setupPageNavigationListeners twice.
+      // Core's store setup is not idempotent.
+      document.removeEventListener('DOMContentLoaded', initialPageLoadHandler);
+      document.removeEventListener('readystatechange', initialPageLoadHandler);
+      setupPageNavigationListeners();
+    };
+
+    document.addEventListener('DOMContentLoaded', initialPageLoadHandler);
+
+    if (document.readyState === 'interactive') {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts:94]: Duplicate of https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3088619069 — rationale already in the block comment at `pageLifecycle.ts:70-81`.

File: packages/react-on-rails/src/clientStartup.ts [hunk 15]
```diff
@@ -1,31 +1,46 @@
...
   globalThis.__REACT_ON_RAILS_EVENT_HANDLERS_RAN_ONCE__ = true;
 
   // Core package: Wait for full page load, then render all components
   // Pro package: Can start hydration immediately or wait for page load
-  onPageLoaded(reactOnRailsPageLoaded);
+  onPageLoaded(runAutomaticPageLoad);
+  return true;
+}
+
+export function reactOnRailsPageLoaded() {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/clientStartup.ts]: Declining. The premise is factually incorrect — `clientStartup()` in the same file does not have an explicit return type either (it returns void, not boolean). Both exported functions are already consistent with each other. Additionally, `clientStartup.ts` isn't modified by this PR, so adding a type annotation here would be out of scope.

**Comment by ihabadham** [comment]
Re: CodeRabbit nitpick on `clientStartup.test.js` setOptions timing (in https://github.com/shakacode/react_on_rails/pull/3151#pullrequestreview-4116011491): `clientStartup.test.js` isn't modified by this PR — branch scope is `CHANGELOG.md`, `pageLifecycle.ts`, and `pageLifecycle.test.js`. Skipping.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: full PR history (no prior summary comment).

### Mattered
- None. 0 MUST-FIX and 0 DISCUSS items.

### Skipped (23 total: 22 inline + 1 review-summary nitpick)
- **5 items against reverted/simplified code** (commits `05e43fb3a` reverted `219545c9c`; `3c3ebcf07` simplified handler setup): `consumeInitialPageLoadIfNeeded`, `startupWasPending`, `cleanupListeners` TDZ, nav-listener-during-recovery asymmetry, deferred `setOptions` timing — all reference functions/variables that no longer exist.
- **2 items about `clientStartup.test.js`** (not modified by this PR): test name rename, cleanup assertion in integration test.
- **6 test-style preferences / duplicates**: negative assertion for loading state (x2), `findLast` vs `find` in test helper (x3), test for defensive readystatechange-while-still-interactive filter.
- **5 items we explicitly decided against this session**: `isNavigationListenersSetUp` defensive flag; guard for spurious `removeEventListener` (harmless no-op); test rewrite with `mockRestore` (harness doesn't support it); mechanism-level `removeEventListener` assertions on recovery test (x2, same rationale as dropping the analogous test in `2a655fd19`).
- **3 doc-nit duplicates**: readystatechange-during-interactive asymmetry comment — rationale already in block comment at `pageLifecycle.ts:70-81`.
- **1 return-type annotation request** with factually-wrong premise: `clientStartup()` also has no explicit return type, so the claimed inconsistency doesn't exist; file isn't in this PR either.
- **1 CodeRabbit review-summary nitpick** about `clientStartup.test.js` setOptions timing (also out of scope).

All 22 inline review threads have rationale replies and are resolved. CodeRabbit review summary nitpick has a standalone issue-comment reply.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 16]
```diff
@@ -67,20 +67,33 @@ function initializePageEventListeners(): void {
...
+  //
+  // ReactOnRailsPro's early hydration path is more resilient to the registration race because it can
+  // hydrate components as their HTML and bundles arrive, but this page lifecycle still powers the
+  // fallback page-load sweep. See pageLifecycle.test.js for regression coverage of both cases.
   if (document.readyState === 'complete') {
     setupPageNavigationListeners();
   } else {
-    document.addEventListener('DOMContentLoaded', setupPageNavigationListeners);
+    const initialPageLoadHandler = (event: Event): void => {
+      if (event.type === 'readystatechange' && document.readyState !== 'complete') return;
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts:86]: Declining. The filter line is defensive insurance against synthetic events, test dispatch, and `document.open()` edge cases. In real browsers the `readystatechange` listener only fires at the `complete` transition from our registration point: we register only when `readyState === 'interactive'`, and `readyState` progresses monotonically. Adding a test would pin behavior that the browser can't naturally produce — same category of over-specification we've been avoiding throughout this PR.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 17]
```diff
@@ -67,20 +67,33 @@ function initializePageEventListeners(): void {
...
+      document.removeEventListener('DOMContentLoaded', initialPageLoadHandler);
+      document.removeEventListener('readystatechange', initialPageLoadHandler);
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts:88]: Declining — same reasoning as why the analogous cleanup test was dropped in `2a655fd19`. Asserting `removeEventListener` was called is a mechanism-level assertion, not a behavioral one. It would fail on valid refactors (e.g. AbortController) that preserve the cleanup invariant via a different mechanism, without catching cleanup bugs that leave the calls in place but misapply them. The invariant itself (`setupPageNavigationListeners` runs at most once) can't be tested behaviorally in this harness: the spy mocks `addEventListener` with a no-op, so tests retrieve handlers via the spy's captured calls — cleanup doesn't actually prevent those captured handler references from being invoked.

File: packages/react-on-rails/src/pageLifecycle.ts [hunk 18]
```diff
@@ -67,20 +67,33 @@ function initializePageEventListeners(): void {
...
+      setupPageNavigationListeners();
+    };
+
+    document.addEventListener('DOMContentLoaded', initialPageLoadHandler);
+
+    if (document.readyState === 'interactive') {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/src/pageLifecycle.ts:94]: The rationale for the asymmetry is already documented in the block comment at `pageLifecycle.ts:70-81`, which covers why the `interactive` path specifically needs the `complete` fallback. Adding another comment at the registration site or a negative assertion in the `loading` test would duplicate coverage of the same asymmetry.

File: packages/react-on-rails/tests/pageLifecycle.test.js [hunk 19]
```diff
@@ -31,6 +31,9 @@ describe('pageLifecycle', () => {
   // eslint-disable-next-line global-require
   const importPageLifecycle = () => require('../src/pageLifecycle.ts');
 
+  const getEventHandler = (eventName) =>
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails/tests/pageLifecycle.test.js:34]: Duplicate of https://github.com/shakacode/react_on_rails/pull/3151#discussion_r3089107610 — declined for the same reason. Tests in this file don't register the same event name multiple times and modules reset between tests. If that invariant changes later, the helper can be updated alongside the test that introduces the multi-registration pattern.
